### PR TITLE
feat(dd): add traces to gitfilesysteM

### PR DESCRIPTION
**Pull request by kishore03109**
## Problem

we are having divergence, still not enough information to determine why. adding traces to the major gitfile system operations to give a better idea of what is happening. 

## Notes to all dev

Once this PR is merged all devs will need to call npm install to:
1. get the updated version of [dd-trace](https://www.npmjs.com/package/dd-trace)
2. ensure that [patch-package](https://www.npmjs.com/package/patch-package) runs to patch dd-trace and make it compatible with [neverthrow](https://github.com/supermacro/neverthrow)

## Deploy Notes

This PR records a lot of new spans into traces. Basically ALL `GitFileSystem` operations are now instrumented.
During release, close attention need to be given to system load to ensure the new instrumentation is not adding too high a CPU cost to the system.

### CODE REVIEW

**Comment by kishore03109** [comment]
* **#1240** <a href="https://app.graphite.dev/github/pr/isomerpages/isomercms-backend/1240?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @kishore03109 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by timotheeg** [comment]
* Added tests to document the compatibility behaviour of dd-trace and neverthrow
* Upgraded dd-trace to v5 (we should check if our docker uses the latest dd-agent too...)

**Comment by timotheeg** [comment]
Test on staging with dd-trace V5, sample [trace here](https://ogp.datadoghq.com/apm/trace/2311127660494155708?colorBy=service&env=staging&graphType=flamegraph&shouldShowLegend=true&sort=time&spanID=1716086828328783162&spanViewType=databases&timeHint=1711504205436&view=spans) for reference

**Review by timotheeg — APPROVED** [review]
LGTM - tested on staging

Warning on potential cost:
* runtime cpu cost since we now instrument ALL the git actions
* potential monetary cost if we index a lot more spans

**Comment by seaerchin** [comment]
@timotheeg @kishore03109 we seem to have git patches here? not sure if this was intended

**Comment by timotheeg** [comment]
> we seem to have git patches here? not sure if this was intended

It's intended. We have to patch `dd-trace` to make it compatible with neverthrow. This is done with using [patch-package](https://www.npmjs.com/package/patch-package)

**Comment by timotheeg** [comment]
* Removed one of the patch for `dd-trace@4.26`, since we're locking the version to 5.9 now
* Added blacklist of methods to instrument, to reduce span noise just a little

File: patches/dd-trace+5.9.0.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
+diff --git a/node_modules/dd-trace/packages/dd-trace/src/tracer.js b/node_modules/dd-trace/packages/dd-trace/src/tracer.js
+index 63c60e8..49b5b7e 100644
+--- a/node_modules/dd-trace/packages/dd-trace/src/tracer.js
++++ b/node_modules/dd-trace/packages/dd-trace/src/tracer.js
+@@ -73,7 +73,19 @@ class DatadogTracer extends Tracer {
+ 
+       const result = this.scope().activate(span, () => fn(span))
+ 
+-      if (result && typeof result.then === 'function') {
```
Comment (seaerchin) [inline patches/dd-trace+5.9.0.patch:9]: wouldn't we lose compat with promises in this case?

Comment (seaerchin) [reply patches/dd-trace+5.9.0.patch:9]: oops, saw the extra stuff below

**Pull request merged**